### PR TITLE
[monodroid] support typemaps from shared runtime again

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/xamarin/xamarin-android/commit/7ac95e733c6065dae0c65cad217ab8214fd72247
Context: https://github.com/xamarin/xamarin-android/commit/decfbccf#diff-fcb50a6533eb26323059bc5c8bda1dd7L384

This reverts 7ac95e7 and reverts a small portion of decfbcc.

Unfortunately with 16.2/master we have a ~2.75 second regression with
incremental build performance.

An example using a "Hello World" Xamarin.Forms app, which came from
the File->New Project template in Visual Studio 2019 16.0:

    9.2.3-0
       829 ms  GenerateJavaStubs                          1 calls
    9.4.0-17
      3703 ms  GenerateJavaStubs                          1 calls

This was `Debug` / fast dev build after a XAML change.

The slowdown is due to changes in 7ac95e7, which fixed a crash on
startup when the Mono shared runtime is used.

This constructor used in `<GenerateJavaStubs/>`:

    new TypeNameMapGenerator (ResolvedAssemblies.Select (p => p.ItemSpec), logger)

Will iterate over all the types in the assemblies again, and not make
use of the local `types` list already present.

We removed support in decfbcc for loading typemap files from the Mono
shared runtime. For getting something we can easily cherry-pick for
16.2, I think we can just put this support back in--and completely
revert 7ac95e7.

Down the road we can do something more involved to put back the
changes from decfbcc.

### CODE REVIEW

File: src/monodroid/jni/embedded-assemblies.cc [hunk 1]
```diff
@@ -400,6 +392,17 @@ EmbeddedAssemblies::gather_bundled_assemblies_from_apk (const char* apk, monodro
...
+				md_mmap_info map_info   = md_mmap_apk_file (fd, offset, info.uncompressed_size, cur_entry_name, apk);
+				add_type_mapping (&java_to_managed_maps, apk, cur_entry_name, (const char*)map_info.area);
+				continue;
+			}
+			if (utils.ends_with (cur_entry_name, ".mj")) {
+				md_mmap_info map_info   = md_mmap_apk_file (fd, offset, info.uncompressed_size, cur_entry_name, apk);
+				add_type_mapping (&managed_to_java_maps, apk, cur_entry_name, (const char*)map_info.area);
+				continue;
+			}
+
```
Comment (jonathanpeppers) [inline src/monodroid/jni/embedded-assemblies.cc]: @grendello is there something I can do to skip these two blocks when the app starts with the shared runtime being used?

File: src/monodroid/jni/embedded-assemblies.cc [hunk 2]
```diff
@@ -147,15 +143,13 @@ EmbeddedAssemblies::find_entry_in_type_map (const char *name, uint8_t map[], Typ
...
 {
-#if defined (DEBUG) || !defined (ANDROID)
 	for (TypeMappingInfo *info = java_to_managed_maps; info != nullptr; info = info->next) {
 		/* log_warn (LOG_DEFAULT, "# jonp: checking file: %s!%s for type '%s'", info->source_apk, info->source_entry, java); */
 		const char *e = reinterpret_cast<const char*> (bsearch (java, info->mapping, info->num_entries, info->entry_length, TypeMappingInfo_compare_key));
 		if (e == nullptr)
 			continue;
 		return e + info->value_offset;
 	}
-#endif
```
Comment (jonathanpeppers) [inline src/monodroid/jni/embedded-assemblies.cc]: @grendello same here, I could skip this for-loop unless the shared runtime is used.

**Comment by jonathanpeppers** [comment]
Here are the binlogs showing the time difference in `<GenerateJavaStubs/>`: [HelloWorld.zip](https://github.com/xamarin/xamarin-android/files/3234035/HelloWorld.zip)

File: src/monodroid/jni/embedded-assemblies.cc (hunk 1 above)
Comment (grendello) [reply src/monodroid/jni/embedded-assemblies.cc]: Yep, we can [add a flag here](https://github.com/xamarin/xamarin-android/blob/9887642623f6cea964e8cfe302bca83f6387d068/src/monodroid/jni/xamarin-app.h#L17), initialize it to some [default here](https://github.com/xamarin/xamarin-android/blob/9887642623f6cea964e8cfe302bca83f6387d068/src/monodroid/jni/application_dso_stub.cc#L15), add an appropriate [property here](https://github.com/xamarin/xamarin-android/blob/9887642623f6cea964e8cfe302bca83f6387d068/src/Xamarin.Android.Build.Tasks/Utilities/ApplicationConfigNativeAssemblyGenerator.cs#L7), write it out to the [assembly source here](https://github.com/xamarin/xamarin-android/blob/9887642623f6cea964e8cfe302bca83f6387d068/src/Xamarin.Android.Build.Tasks/Utilities/ApplicationConfigNativeAssemblyGenerator.cs#L42-L70) (the fields **must** be written in the same order they are declared in the `xamarin-app.h` header!), assign the [correct value here](https://github.com/xamarin/xamarin-android/blob/9887642623f6cea964e8cfe302bca83f6387d068/src/Xamarin.Android.Build.Tasks/Tasks/GeneratePackageManagerJava.cs#L259-L267) and then query it like so whenever you need to know its value:

```cpp
if (application_config.YourFlagName)
  ;
```

When adding the flag, please keep it close to the other boolean flags :)

File: src/monodroid/jni/embedded-assemblies.cc (hunk 2 above)
Comment (grendello) [reply src/monodroid/jni/embedded-assemblies.cc]: ditto :)

**Comment by dellis1972** [comment]
@jonathanpeppers we need to back out this change as well https://github.com/xamarin/xamarin-android/pull/3084/files

**Comment by dellis1972** [comment]
@jonathanpeppers ok ignore me, you already did lol

**Comment by jonathanpeppers** [comment]
I think the macOS test failures are OK:
```
(2 failures / +2)
Test collection for System.Threading.ThreadPools.Tests.ThreadPoolTests.System.Threading.ThreadPools.Tests.ThreadPoolTests.QueueRegisterPositiveAndFlowTest / Release
Test collection for System.Threading.ThreadPools.Tests.ThreadPoolTests.System.Threading.ThreadPools.Tests.ThreadPoolTests.QueueRegisterPositiveAndFlowTest / Release
```

I think these are happening on master.

**Pull request merged**